### ISSUE TITLE
Can activation_checkpointing offloads to NVMe?

### ISSUE THREAD

**Issue by ghosthamlet** [microsoft/DeepSpeed#1201]
There is a `cpu_checkpointing` in config, why can't offload it to NVMe?

**Comment by tjruwase** [comment]
@ghosthamlet, thanks for the question. Yes, activations can also be offloaded to NVMe. We have not had bandwidth to implement this feature.

**Comment by ghosthamlet** [comment]
@tjruwase are there a plan to implement it? I think the activations are very large for some model, cpu_checkpointing is not enough.

**Comment by tjruwase** [comment]
Yes, this is in our plans to implement. Can you share some details about your scenario for our planning? Thanks!

**Comment by ghosthamlet** [comment]
@tjruwase My scenario is simple, i want to train a GPT-2 model with params larger than 2.7B (the model config is same as GPT-3 2.7B), seq_len (context window) is 1024 tokens, and batch_size on single gpu should larger than 12, the datasets is IterableDataset.
I have two 2080ti GPU, 128G cpu memory, 2T NVMe, 
so i think if activation_checkpointing can't offload to NVMe, the 128G memory maybe not enough, do you think so? Thanks.

**Comment by tjruwase** [comment]
Thanks for sharing your scenario. However, I think if you offload optimizer to NVMe, you should have enough cpu memory for activations for a 2.7B model. Can you share the hidden size and layer count as well?

**Comment by ghosthamlet** [comment]
Thanks for your fast reply, the model config is: layers 32, hidden size 2560, heads 32, head size 80.

**Comment by tjruwase** [comment]
@ghosthamlet, thanks for sharing these details. By my estimation, activation memory for batch size of 10 per gpu should be ~4GB. Are you able to test this out in practice, perhaps there are errors in my calculations :).  With 64G memory, you might even be able to keep optimizer in cpu. Please let me know how it goes.

**Comment by ghosthamlet** [comment]
I will give results here, maybe next week i will try it. Thank you very much for all the patiently answers.

**Comment by ghosthamlet** [comment]
@tjruwase I have succeed in training 2.6B GPT-2 model,  i have to finetune on a existing 2.6B model, so did not train 2.7B model, but the difference should not be large. 
Yes, your estimate is right, no need NVMe for 2.6B model, it just used 53.17G cpu memory.

**hardware:** 
two 2080ti GPUs (each 11G memory), 64G cpu memory.

| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |

|100%   71C    P2   103W / 250W |  10474MiB / 11019MiB |      5%      Default |
+-------------------------------+----------------------+----------------------+

| 62%   72C    P2   111W / 250W |   8447MiB / 11019MiB |      0%      Default |
+-------------------------------+----------------------+----------------------+


**model config:** 
```
{
  "n_ctx": 1024,  // seq_len (context window)
  "n_embd": 2560, // hidden size
  "n_head": 32,
  "n_layer": 32,
  "vocab_size": 30000
}
```

**deepspeed config:** 
```
{
 "gradient_clipping": 1.0,
 "zero_optimization": {
  "stage": 3,
  "contiguous_gradients": true,
  "overlap_comm": false,
  "reduce_bucket_size": 50000000,
  "stage3_max_live_parameters": 100.0,
  "stage3_max_reuse_distance": 100.0,
  "stage3_prefetch_bucket_size": 100.0,
  "stage3_param_persistence_threshold": 100.0,
  "sub_group_size": 100.0,
  "offload_optimizer": {
   "device": "cpu"
  },
  "offload_param": {
   "device": "cpu"
  }
 },
 "train_batch_size": 20,
 "gradient_accumulation_steps": 1,
 "fp16": {
  "enabled": true,
  "loss_scale": 0,
  "loss_scale_window": 1000,
  "hysteresis": 2,
  "min_loss_scale": 1
 },
 "activation_checkpointing": {
  "partition_activations": true,
  "num_checkpoints": 1,
  "cpu_checkpointing": true
 },
 "wall_clock_breakdown": false,
}
```

But i just can train with batch_size 10 per gpu. when train with batch_size 12, the 11G GPU memory is not enough when saving fp16 model.

**Comment by tjruwase** [comment]
@ghosthamlet, can you explain what you mean by "when saving fp16 model"? I am confused whether the OOM happens during training or during checkpoint saving?

To train with large micro-batch size can you try using gradient accumulation steps > 1? This will reduce the instantaneous micro batch size and activations memory consumption.

**Comment by ghosthamlet** [comment]
@tjruwase Thanks for the reply. 
The mean of "when saving fp16 model" is call `deepspeed_engine.save_fp16_model`, in DeepSpeedEngine code `save_fp16_model` called `_zero3_consolidated_fp16_state_dict`, comment says: consolidation is expensive in time and memory and therefore isn't a default, so i think OOM is normal in my config.

Yes, using gradient accumulation steps > 1 is working, the above OOM problem can be avoid.
Thanks, i can close this issue now.

**Comment by tjruwase** [comment]
@ghosthamlet, I am glad that things are working now. However, I am going to keep this issue open to address the original request.

**Comment by ghosthamlet** [comment]
@tjruwase Thanks. I'm looking forward to that.
DeepSpeed is an incredible library, never thought i can train 2.6B params model on an 2080Ti GPU, and with fast enough speed. 
I leaned much from this library, its articles and papers. Thanks again.

Do you have some roadmap for DeepSpeed project? Like Some new features plan, or some big refactor, like what Transformers(https://github.com/huggingface/transformers/) did (they modularized their single huge classes/files).
